1354 - Add support for adding and removing a project from Dataset

nozomione opened the pull request:
## Issue Number

Closes #1354 

Stacked PR of #1374 

## Purpose/Implementation Notes

I've added support for adding and removing projects from a dataset on the UI.

@dvenprasad, see the latest UI [here](https://scpca-portal-qkm3c7c1b-ccdl.vercel.app/). Before adding projects to a dataset, ensure you generate a token  (this authorization is temporary  and will be removed later). Thank you!

### Changes include:

- In `DatasetAddProjectModal`:
    - Added states for managing project options (i.e., `modalities`, `excludeMultiplexed`, `includeBulk`, `includeMerge` ) used in the corresponding sub-components
    - Added states for populating the project data for `myDataset` (`dataset.data[project.scpca_id]`) used by `addProject` (e.g., `projectData`)
- Each sub-components (for project options) sets the corresponding parent states 

- In `useDatasetManager` :
  -  Adjusted `addProject`, `createDataset`,  and `updateDataset`
  -  Added new methods:
    -  `getAddedProjectModalities`, `isProjectAddedToDataset`, for the modal presentation
    - `getProjectDataSamples`, `getProjectSingleCellSamples`, and `getProjectSpatialSamples` for populating the project data for `addDataset` (replacing `getProjectData`)

- Added a new helper `differenceArray`  for calculating sample differences (also used in Stacked PRs) 

### Currently implemented flows via `Add Project to Dataset` modal:

 - Add a project to `myDataest` by selecting project options, then click the `Add to Dataset` button.

Once `myDataset` has been created:
- Replace the `"Add to Dataset" button ` with the `Added to Dataset` text  (see [comment](https://github.com/AlexsLemonade/scpca-portal/pull/1376#issuecomment-3176327662))

**NOTE:** A new issue will be filed for preselecting the project options' checkboxes based on the most recently added project using localStorage.

## Types of changes

- Refactor (addresses code organization and design mentioned in corresponding issue)
- New feature (non-breaking change which adds functionality)

## Functional tests
> [!note]
> All testing was performed on `localhost`. Switching to `staging` API should work once it’s available - any necessary updates will be adjusted accordingly.
>  (For dataset creation, a token was temporarily generated using Postman for `createDataset` request, as it currently requires authentication. However, the authentication will be removed in a future BE update.)

####  UI Conditions and checks tested:
- Replace the `"Add to Dataset" button ` with the `Added to Dataset` text on the project card
- In the modal:
  - Disable the format option once `myDataset` has been created
  - (Subject to change) Disable the he `SPATIAL` modality checkbox if `ANN_DATA` format is selected
  - Disable the merged object checkbox in the following cases:
    -  When only the `SPATIAL` modality checkbox is selected
    -  When a project contains multiplexed samples. Instead display the `“Merged objects are not available for projects with multiplexed samples.”` warning message


## Checklist

<!-- Put an `x` in the boxes that apply. -->

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

<img width="951" height="91" alt="datasets - historical_records" src="https://github.com/user-attachments/assets/19abd36e-4c5e-45ee-a8c6-b663226e74a7" />

<img width="944" height="102" alt="my-dataset" src="https://github.com/user-attachments/assets/2f0594d9-9779-4620-ac88-1ce40ff56fb8" />

davidsmejia commented on client/src/components/DatasetProjectAdditionalOptions.js:
```diff
@@ -1,22 +1,35 @@
+export const DatasetProjectAdditionalOptions = ({
+  project,
+  format,
+  handleIncludeBulk,
+  handleIncludeMerge
```
```suggestion
  onIncludeBulkChange,
  onIncludeMergeChange
```

davidsmejia commented on client/src/components/DatasetProjectDataFormat.js:
```diff
@@ -1,14 +1,24 @@
-  const [format, setFormat] = useState(formatOptions[0])
+export const DatasetProjectDataFormat = ({
+  project,
+  format,
+  handleSetFormat
```
```suggestion
  onFormatChange
```

nozomione replied:
🗒️ We've decided to define the `format` state within this component (instead of the parent) and no longer require the `onFormatChange` prop.

davidsmejia commented on client/src/components/DatasetProjectDataFormat.js:
```diff
@@ -19,11 +29,12 @@ export const DatasetProjectDataFormat = () => {
         valueKey={{ key: 'value', reduce: true }}
-        value={format || formatOptions[0]}
-        onChange={({ value }) => setFormat(value)}
+        value={myDataset.format || format}
```
```suggestion
        value={format}
```

davidsmejia commented on client/src/components/DatasetProjectDataFormat.js:
```diff
@@ -1,14 +1,24 @@
+  format,
+  handleSetFormat
+}) => {
+  const { myDataset } = useDatasetManager()
+
+  const defaultOptions = [
+    { key: 'SINGLE_CELL_EXPERIMENT', value: project.has_single_cell_data },
+    { key: 'ANN_DATA', value: project.includes_anndata }
+  ]
+    .filter((f) => f.value)
+    .map((f) => f.key)
```
```suggestion
  const { myDataset, userFormat } = useDatasetManager()

  const defaultOptions = [
    { key: 'SINGLE_CELL_EXPERIMENT', value: project.has_single_cell_data },
    { key: 'ANN_DATA', value: project.includes_anndata }
  ]
    .filter((f) => f.value)
    .map((f) => f.key)
  
  const defaultFormat = defaultOptions.includes(userFormat) ? userFormat : defaultOptions[1]
  
  const [format, setFormat] = useState(myDatset.format || defaultFormat)
  
  const handleFormatChange = (newFormat) => {
    setFormat(newFormat)
    onFormatChange(newFormat)
  }
  
```

nozomione commented:
I've applied the feedback and also made the following updates:
-  Disable the merged object checkbox when only the `SPATIAL` modality checkbox is selected
-  Ensured that the “Merged objects are not available for projects with multiplexed samples.” warning message is display only if a project contains multiplexed samples
- Reset all parent states on modal close, so reopening the modal or new form submissions won’t be affected by previously selected user values
- Added the `excludeMultiplexed` state to the `DatasetAddProjectModal` parent component for filtering multiplexed samples, and updated getProjectData accordingly with a `TODO` comment (temporarily using the `countArrayElements` helper - newly added)

This PR is ready for another look. Thank you, David!

avrohomgottlieb reviewed (approved):
Great PR!

davidsmejia commented on client/src/components/DatasetProjectAdditionalOptions.js:
```diff
@@ -1,36 +1,90 @@
 
-  const isMergedObjectsAvailable = includesMergedSce || includesMergedAnnData
-  const isExcludeMultiplexedAvailable = false
+  const isMergedObjectsAvailable =
+    (includesMergedSce || includesMergedAnnData) &&
+    (selectedModalities.includes('SINGLE_CELL') ||
+      (selectedModalities.length === 0 && hasSingleCellData && !hasMultiplexed))
```
I dont think we need to do this much logic here.

1. Depending on the format of the dataset we can check the appropriate `includes_merged_sce` or `includes_merged_ann_data`.
2. If the user has no format you can check either of them.
3. `hasMultiplexed` and `hasSingleCellData` do not need to be checked here. that information is contained in `includes_merged_sce` etc.
4. You only need to check `hasMultiplexed` to show the explanation when merged is not available.

nozomione commented:
I've applied your feedback and updated the Purpose/Implementation Notes section to reflect these changes. This PR is ready for another look. Thank you, David!

davidsmejia reviewed (approved):
LGTM
